common/predictive-text/promise-store.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2018 National Research Council Canada (author: Eddie A. Santos)
+ * Copyright (c) 2018 SIL International
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+
+type Resolve<T> = (value?: T | PromiseLike<T>) => void;
+type Reject = (reason?: any) => void;
+interface PromiseCallbacks<T> {
+  resolve: Resolve<T>;
+  reject: Reject;
+}
+
+/**
+ * Associate tokens with promises.
+ *
+ * First, .make() a promise -- associate a token with resolve/reject callbacks.
+ *
+ * You can either .keep() a promise -- resolve() and forget it;
+ * Or you may also .break() a promise -- reject() and forget it.
+ * 
+ * <T> is the type of resolved value (value yielded successfully by promise).
+ */
+class PromiseStore<T> {
+  private _promises: Map<Token, PromiseCallbacks<T>>;
+  constructor() {
+    this._promises = new Map();
```

Comment:
This is an ES6 feature, but we only require "basic" support `.set()`, `.get()`, `.has()` and `.delete()`. IE11 supports all of these <http://kangax.github.io/compat-table/es6/#test-Map>.